File: src/main/resources/kieprocess-js/process_service.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module kieprocess-js/process_service */
var utils = require('vertx-js/util/utils');
var Vertx = require('vertx-js/vertx');
var ProcessInstanceService = require('kieprocess-js/process_instance_service');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JProcessService = com.diabolicallabs.process.manager.service.ProcessService;

/**
 @class
*/
var ProcessService = function(j_val) {

  var j_processService = j_val;
  var that = this;

  /**

   @public
   @param processInstanceId {number} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.abort = function(processInstanceId, handler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] ==='number' && typeof __args[1] === 'function') {
      j_processService["abort(java.lang.Long,io.vertx.core.Handler)"](utils.convParamLong(processInstanceId), function(ar) {
      if (ar.succeeded()) {
        handler(null, null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param processId {string} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.create = function(processId, handler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] === 'string' && typeof __args[1] === 'function') {
      j_processService["create(java.lang.String,io.vertx.core.Handler)"](processId, function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), ProcessInstanceService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param processId {string} 
   @param variables {Object} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.createWithVariables = function(processId, variables, handler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function') {
      j_processService["createWithVariables(java.lang.String,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](processId, utils.convParamJsonObject(variables), function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), ProcessInstanceService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param eventName {string} 
   @param data {Object} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.signalEvent = function(eventName, data, handler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function') {
      j_processService["signalEvent(java.lang.String,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](eventName, utils.convParamJsonObject(data), function(ar) {
      if (ar.succeeded()) {
        handler(null, null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param eventName {string} 
   @param processInstanceId {number} 
   @param data {Object} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.signalEventForProcess = function(eventName, processInstanceId, data, handler) {
    var __args = arguments;
    if (__args.length === 4 && typeof __args[0] === 'string' && typeof __args[1] ==='number' && (typeof __args[2] === 'object' && __args[2] != null) && typeof __args[3] === 'function') {
      j_processService["signalEventForProcess(java.lang.String,java.lang.Long,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](eventName, utils.convParamLong(processInstanceId), utils.convParamJsonObject(data), function(ar) {
      if (ar.succeeded()) {
        handler(null, null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param processId {string} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.startProcess = function(processId, handler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] === 'string' && typeof __args[1] === 'function') {
      j_processService["startProcess(java.lang.String,io.vertx.core.Handler)"](processId, function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), ProcessInstanceService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param processId {string} 
   @param jsonObject {Object} 
   @param handler {function} 
   @return {ProcessService}
   */
  this.startProcessWithVariables = function(processId, jsonObject, handler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function') {
      j_processService["startProcessWithVariables(java.lang.String,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](processId, utils.convParamJsonObject(jsonObject), function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), ProcessInstanceService), null);
      } else {
        handler(null, ar.cause());
      }
    });
      return that;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_processService;
};

/**

 @memberof module:kieprocess-js/process_service
 @param vertx {Vertx} 
 @param address {string} 
 @return {ProcessService}
 */
ProcessService.createProxy = function(vertx, address) {
  var __args = arguments;
  if (__args.length === 2 && typeof __args[0] === 'object' && __args[0]._jdel && typeof __args[1] === 'string') {
    return utils.convReturnVertxGen(JProcessService["createProxy(io.vertx.core.Vertx,java.lang.String)"](vertx._jdel, address), ProcessService);
  } else throw new TypeError('function invoked with invalid arguments');
};

// We export the Constructor function
module.exports = ProcessService;

File: src/main/ceylon/com/diabolicallabs/process/manager/ceylon/kieprocess/service/ProcessService.java
package com.diabolicallabs.process.manager.ceylon.kieprocess.service;

import com.redhat.ceylon.compiler.java.metadata.Ceylon;
import com.redhat.ceylon.compiler.java.metadata.TypeInfo;
import com.redhat.ceylon.compiler.java.metadata.TypeParameter;
import com.redhat.ceylon.compiler.java.metadata.TypeParameters;
import com.redhat.ceylon.compiler.java.metadata.Variance;
import com.redhat.ceylon.compiler.java.metadata.Ignore;
import com.redhat.ceylon.compiler.java.metadata.Name;
import com.redhat.ceylon.compiler.java.runtime.model.TypeDescriptor;
import com.redhat.ceylon.compiler.java.runtime.model.ReifiedType;
import ceylon.language.Callable;
import ceylon.language.DocAnnotation$annotation$;
import io.vertx.ceylon.core.Vertx;
import io.vertx.core.json.JsonObject;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;

@Ceylon(major = 8)
public class ProcessService implements ReifiedType {

  @Ignore
  public static final io.vertx.lang.ceylon.ConverterFactory<com.diabolicallabs.process.manager.service.ProcessService, ProcessService> TO_CEYLON = new io.vertx.lang.ceylon.ConverterFactory<com.diabolicallabs.process.manager.service.ProcessService, ProcessService>() {
    public io.vertx.lang.ceylon.Converter<com.diabolicallabs.process.manager.service.ProcessService, ProcessService> converter(final TypeDescriptor... descriptors) {
      return new io.vertx.lang.ceylon.Converter<com.diabolicallabs.process.manager.service.ProcessService, ProcessService>() {
        public ProcessService convert(com.diabolicallabs.process.manager.service.ProcessService src) {
          return new ProcessService(src);
        }
      };
    }
  };

  @Ignore
  public static final io.vertx.lang.ceylon.Converter<ProcessService, com.diabolicallabs.process.manager.service.ProcessService> TO_JAVA = new io.vertx.lang.ceylon.Converter<ProcessService, com.diabolicallabs.process.manager.service.ProcessService>() {
    public com.diabolicallabs.process.manager.service.ProcessService convert(ProcessService src) {
      return src.delegate;
    }
  };

  @Ignore public static final TypeDescriptor $TypeDescriptor$ = TypeDescriptor.klass(ProcessService.class);
  @Ignore private final com.diabolicallabs.process.manager.service.ProcessService delegate;

  public ProcessService(com.diabolicallabs.process.manager.service.ProcessService delegate) {
    this.delegate = delegate;
  }

  @Ignore 
  public TypeDescriptor $getType$() {
    return $TypeDescriptor$;
  }

  @Ignore
  public Object getDelegate() {
    return delegate;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService abort(
    final @TypeInfo("ceylon.language::Integer") @Name("processInstanceId")  ceylon.language.Integer processInstanceId, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable?)") @Name("handler")  Callable<?> handler) {
    java.lang.Long arg_0 = io.vertx.lang.ceylon.ToJava.Long.safeConvert(processInstanceId);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> arg_1 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.Void>(handler) {
      public Object toCeylon(java.lang.Void event) {
        return null;
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.abort(arg_0, arg_1));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService create(
    final @TypeInfo("ceylon.language::String") @Name("processId")  ceylon.language.String processId, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessInstanceService)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(processId);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.ProcessInstanceService>> arg_1 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.ProcessInstanceService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.ProcessInstanceService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessInstanceService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.create(arg_0, arg_1));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService createWithVariables(
    final @TypeInfo("ceylon.language::String") @Name("processId")  ceylon.language.String processId, 
    final @TypeInfo("ceylon.json::Object") @Name("variables")  ceylon.json.Object variables, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessInstanceService)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(processId);
    io.vertx.core.json.JsonObject arg_1 = io.vertx.lang.ceylon.ToJava.JsonObject.safeConvert(variables);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.ProcessInstanceService>> arg_2 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.ProcessInstanceService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.ProcessInstanceService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessInstanceService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.createWithVariables(arg_0, arg_1, arg_2));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService signalEvent(
    final @TypeInfo("ceylon.language::String") @Name("eventName")  ceylon.language.String eventName, 
    final @TypeInfo("ceylon.json::Object") @Name("data")  ceylon.json.Object data, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable?)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(eventName);
    io.vertx.core.json.JsonObject arg_1 = io.vertx.lang.ceylon.ToJava.JsonObject.safeConvert(data);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> arg_2 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.Void>(handler) {
      public Object toCeylon(java.lang.Void event) {
        return null;
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.signalEvent(arg_0, arg_1, arg_2));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService signalEventForProcess(
    final @TypeInfo("ceylon.language::String") @Name("eventName")  ceylon.language.String eventName, 
    final @TypeInfo("ceylon.language::Integer") @Name("processInstanceId")  ceylon.language.Integer processInstanceId, 
    final @TypeInfo("ceylon.json::Object") @Name("data")  ceylon.json.Object data, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable?)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(eventName);
    java.lang.Long arg_1 = io.vertx.lang.ceylon.ToJava.Long.safeConvert(processInstanceId);
    io.vertx.core.json.JsonObject arg_2 = io.vertx.lang.ceylon.ToJava.JsonObject.safeConvert(data);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> arg_3 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<java.lang.Void>(handler) {
      public Object toCeylon(java.lang.Void event) {
        return null;
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.signalEventForProcess(arg_0, arg_1, arg_2, arg_3));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService startProcess(
    final @TypeInfo("ceylon.language::String") @Name("processId")  ceylon.language.String processId, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessInstanceService)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(processId);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.ProcessInstanceService>> arg_1 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.ProcessInstanceService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.ProcessInstanceService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessInstanceService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.startProcess(arg_0, arg_1));
    return this;
  }

  @TypeInfo("com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessService")
  public ProcessService startProcessWithVariables(
    final @TypeInfo("ceylon.language::String") @Name("processId")  ceylon.language.String processId, 
    final @TypeInfo("ceylon.json::Object") @Name("jsonObject")  ceylon.json.Object jsonObject, 
    final @TypeInfo("ceylon.language::Anything(ceylon.language::Throwable|com.diabolicallabs.process.manager.ceylon.kieprocess.service::ProcessInstanceService)") @Name("handler")  Callable<?> handler) {
    java.lang.String arg_0 = io.vertx.lang.ceylon.ToJava.String.safeConvert(processId);
    io.vertx.core.json.JsonObject arg_1 = io.vertx.lang.ceylon.ToJava.JsonObject.safeConvert(jsonObject);
    io.vertx.core.Handler<io.vertx.core.AsyncResult<com.diabolicallabs.process.manager.service.ProcessInstanceService>> arg_2 = handler == null ? null : new io.vertx.lang.ceylon.CallableAsyncResultHandler<com.diabolicallabs.process.manager.service.ProcessInstanceService>(handler) {
      public Object toCeylon(com.diabolicallabs.process.manager.service.ProcessInstanceService event) {
        return com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessInstanceService.TO_CEYLON.converter().safeConvert(event);
      }
    };
    ProcessService ret = com.diabolicallabs.process.manager.ceylon.kieprocess.service.ProcessService.TO_CEYLON.converter().safeConvert(delegate.startProcessWithVariables(arg_0, arg_1, arg_2));
    return this;
  }

}
